### PR TITLE
add mutual friends helper method

### PR DESCRIPTION
**Pull request by nyatindopatrick**
Created mutual friends helper method to check if a user has mutual friendship with other users.

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive work @nyatindopatrick :clap: However, some changes need to be applied in order to get approval for this milestone. Check :arrow_down: 

* [x] when friendship is confirmed, an additional row with reversed user-friend pair is created
  - Confirming the friendship happens on the friendship model level not inside the user model.

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```
#### Your DB associations are working :100: However, you used methods with plain ruby (i.e. not SQL or Active record) syntax to create these associations. Below, I list suggestions on how to do it properly.

* [x] thanks to this, we can find user friends only by user_id in friendships table

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

* [x] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

* [x] User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

* [x] in order to get all posts written by user on their friends, you can use simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
- **[REQUIRED]** Add unit and integration tests related to these associations.

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @nyatindopatrick, @mbabaliiryn :clap: - Good work :+1:, but we still need to fix somethings:**

:x: In the last review, you were asked to move confirm_friend to the friendship model... It is more appropriate there. You will have to review how the friendship controller update method is implemented for this to work.

:x: When a friendship is confirmed, instead of an additional friendship (reverse) to be added to the table, a friendship by the friend with himself is added. Please check this in rails console and fix it.

image

:x: You haven't written appropriate tests for your friendship controller. It is advisable to write integration tests instead. You can see this [guide](https://www.codewithjason.com/rails-integration-tests-rspec-capybara/) on writing integration tests.

--------
**Optional:** You should write model and integrations tests for all features that has been implemented in previous milestones.

_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

File: app/models/user.rb
Comment (whiz25) [inline app/models/user.rb]: - move this to the friendship model

File: app/models/user.rb
Comment (whiz25) [inline app/models/user.rb]: - add tests for these methods

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :clap: 
You have done a good job implementing the suggested changes by the previous reviewers. However, we need to make some changes before we can merge this PR. Check my comments for suggestions


### Martin Afani

File: app/models/user.rb
Comment (nyatindopatrick) [reply app/models/user.rb]: The method doesn't work when moved in the friendship model.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @nyatindopatrick     
#### Great Job so far, but there are a few more things required. 
### General
 - [ ] Lets write integration & unit test for : 
- Comments
- Post
- Like

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes Required :muscle: 
Hi there @nyatindopatrick 
Great job in this project :clap: 
I know many changes were requested already, but we still have one bug to fix. Since this bug is related to one of the most important features from a social media app, I can't approve this PR right now I'm really sorry about it.

- [ ] One of the requests made in previous reviews were not implemented. As it said in the requirements, an additional row needs to be created when a user confirms a Friendship. This feature will make it easy to query relevant data for your application, like querying data for the timeline, which it's not working properly right now. Check out this [article](https://explainextended.com/2009/03/07/selecting-friends/). I created 2 users and made them friends. For each user, I created a post. But only one user was able to see the post from his friend. The other was not able to see it. 
Screenshot from 2020-06-15 10-24-03
Screenshot from 2020-06-15 10-24-21

Please kindly submit for code review again after making these changes.
Happy coding :computer: 
**Sérgio Torres**

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

First, thank you for the improvements done
However, there is more to do to move forward

1. After creating an user, I create its first post but nothing was displayed 

```rb
2.7.0 :001 > Post.all
 Post Load (0.3ms)  SELECT  "posts".* FROM "posts" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Post id: 1, user_id: 1, content: "Alice is there. And  you?", created_at: "2020-06-16 04:07:30", updated_at: "2020-06-16 04:07:30">]> 
2.7.0 :002 > 
```
But as you could see, it's been saved.

2. You didn't fix the problem mentioned by @Torres-ssf  and others. The issues concerning two-row friendships
Here is what I have in the database for two users
```rb
2.7.0 :001 > Friendship.all
 Friendship Load (0.3ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 1, user_id: 2, friend_id: 1, confirmed: true, created_at: "2020-06-16 04:13:19", updated_at: "2020-06-16 04:13:50">]> 
```

3- From the moment BOB has accepted the friendship request sent by her, ALICE was able to see BOB post but not her ones.
Use this method, it could help you
```rb
def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

Kindly make the changes and submit a new code review request
Thanks

Mael FOSSO

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Changes Required.

### Hi team :wave: 

You have made progress.
However, when I accept a friend request, it does not create a mutual friendship. There is no second row created. There is only one friend in the table.

                        Use this image below as a reference.

Screenshot from 2020-06-16 08-46-19

After implementing the changes submit another form for a code review.


### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Review by suretrust — APPROVED** [review]
Hi team, 👋 

### Great job implementing the suggested changes. 🙌  I salute your resilience. 🙇‍♂️ 💯 Let's get this merged! 👍 

_Good luck with your future milestones and happy coding!_ 🎉

**Pull request merged**
